File: cpp/src/SysInfoParsers.cpp

#include "StringUtilities.h"

#include "SysInfoParsers.h"

#include <string>

#include "NodeSysInfo.hpp" //defines NodeSysInfo


void SysInfoParsers::parse_infoblock_v2(NodeSysInfo &_SysInfo, size_t cpuindex, std::string &sBuffer) {
    std::string sProjectName, sSubsystem, sProjectNumber, sSerialNumber, sSWVersion, sDataVersion, sArtNo;
    std::string sRevision, sDate, sTime;
    uint8_t bSystemInfo;
    sStringExtractFixedLengthBinaryString(sBuffer, 8, sProjectName);
    trim_if(sProjectName, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 4, sSubsystem);
    trim_if(sSubsystem, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 6, sRevision); //"%6u"
    trim_if(sRevision, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 10, sDate); //"%4hu-%2hu-%2hu"
    sStringExtractFixedLengthBinaryString(sBuffer, 8, sTime); //"%2hhu:%2hhu:%2hhu"
    sStringExtractFixedLengthBinaryString(sBuffer, 5, sProjectNumber);
    trim_if(sProjectNumber, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 6, sSerialNumber);
    trim_if(sSerialNumber, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 11, sSWVersion); //"%3hu.%3hu.%3hu"
    sStringExtractFixedLengthBinaryString(sBuffer, 11, sDataVersion); //"%3hu.%3hu.%3hu"
    trim_if(sDataVersion, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sStringExtractFixedLengthBinaryString(sBuffer, 11, sArtNo);
    trim_if(sArtNo, is_any_of(" ,.-_\r\t\n" + std::string(1, '\0')));
    sBuffer >> bSystemInfo;

    if (cpuindex == 0) 
    {
        _SysInfo.get<0>() = {sProjectName, sSubsystem, {sSWVersion}};
        _SysInfo.get<1>() = {sProjectName, sSubsystem, {sSWVersion}};
        _SysInfo.get<2>() = {};
        _SysInfo.get<3>() = {};
        _SysInfo.get<4>() = {};
        _SysInfo.get<5>() = {};
    } else {
        //these are partial information .... only store in the corresponding CPU
        _SysInfo.get(cpuindex) = {sProjectName, sSubsystem, {sSWVersion}};
    }

    if (cpuindex == 0 || cpuindex == 1) {
        check_utf8_MakeHexIfBinary(sProjectNumber); //MUST BE VALID UTF8!
        _SysInfo.set<AdditionalInfoKeys::SerialNumber>(sProjectNumber);
        check_utf8_MakeHexIfBinary(sSerialNumber); //MUST BE VALID UTF8!
        _SysInfo.set<AdditionalInfoKeys::ProjectNumber>(sSerialNumber);
        _SysInfo.set<AdditionalInfoKeys::DataVersion>(
            _SysInfo.get(cpuindex).ProjectName_Subsystem() + "_" + sDataVersion);
        check_utf8_MakeHexIfBinary(sArtNo); //MUST BE VALID UTF8!
        _SysInfo.set<AdditionalInfoKeys::ArticleNumber>(sArtNo);
        _SysInfo.set<AdditionalInfoKeys::SystemInfo>((int) bSystemInfo);
    }
}

File: java/src/main/java/org/sammancoaching/SysInfoParsers.java
package org.sammancoaching;

public class SysInfoParsers {
    public static void parse_infoblock_v2(NodeSysInfo _SysInfo, int cpuindex, StringBuilder sBuffer) {
        String sProjectName, sSubsystem, sProjectNumber, sSerialNumber, sSWVersion, sDataVersion, sArtNo;
        String sRevision, sDate, sTime;
        int bSystemInfo;

        sProjectName = extractFixedLength(sBuffer, 8);
        sProjectName = trimIf(sProjectName, " ,.-_\r\t\n\0");

        sSubsystem = extractFixedLength(sBuffer, 4);
        sSubsystem = trimIf(sSubsystem, " ,.-_\r\t\n\0");

        sRevision = extractFixedLength(sBuffer, 6);
        sRevision = trimIf(sRevision, " ,.-_\r\t\n\0");

        sDate = extractFixedLength(sBuffer, 10);
        sTime = extractFixedLength(sBuffer, 8);

        sProjectNumber = extractFixedLength(sBuffer, 5);
        sProjectNumber = trimIf(sProjectNumber, " ,.-_\r\t\n\0");

        sSerialNumber = extractFixedLength(sBuffer, 6);
        sSerialNumber = trimIf(sSerialNumber, " ,.-_\r\t\n\0");

        sSWVersion = extractFixedLength(sBuffer, 11);
        sDataVersion = extractFixedLength(sBuffer, 11);
        sDataVersion = trimIf(sDataVersion, " ,.-_\r\t\n\0");

        sArtNo = extractFixedLength(sBuffer, 11);
        sArtNo = trimIf(sArtNo, " ,.-_\r\t\n\0");

        bSystemInfo = !sBuffer.isEmpty() ? (sBuffer.charAt(0) & 0xFF) : 0;
        if (!sBuffer.isEmpty()) {
            sBuffer.deleteCharAt(0);
        }

        if (cpuindex == 0) {
            _SysInfo.set(0, new CNodeSystemVersion(sProjectName, sSubsystem, new CVersionNumber(sSWVersion)));
            _SysInfo.set(1, new CNodeSystemVersion(sProjectName, sSubsystem, new CVersionNumber(sSWVersion)));
            _SysInfo.set(2, new CNodeSystemVersion());
            _SysInfo.set(3, new CNodeSystemVersion());
            _SysInfo.set(4, new CNodeSystemVersion());
            _SysInfo.set(5, new CNodeSystemVersion());
        } else {
            _SysInfo.set(cpuindex, new CNodeSystemVersion(sProjectName, sSubsystem, new CVersionNumber(sSWVersion)));
        }

        if (cpuindex == 0 || cpuindex == 1) {
            check_utf8_MakeHexIfBinary(sProjectNumber);
            _SysInfo.setAdditionalInfo(AdditionalInfoKeys.SerialNumber, sProjectNumber);
            check_utf8_MakeHexIfBinary(sSerialNumber);
            _SysInfo.setAdditionalInfo(AdditionalInfoKeys.ProjectNumber, sSerialNumber);
            _SysInfo.setAdditionalInfo(AdditionalInfoKeys.DataVersion,
                    _SysInfo.get(cpuindex).getProjectNameSubsystem() + "_" + sDataVersion);
            check_utf8_MakeHexIfBinary(sArtNo);
            _SysInfo.setAdditionalInfo(AdditionalInfoKeys.ArticleNumber, sArtNo);
            _SysInfo.setAdditionalInfo(AdditionalInfoKeys.SystemInfo, bSystemInfo);
        }
    }

    private static String extractFixedLength(StringBuilder buffer, int length) {
        int actualLength = Math.min(length, buffer.length());
        String result = buffer.substring(0, actualLength);
        buffer.delete(0, actualLength);
        return result;
    }

    private static String trimIf(String s, String set) {
        int start = 0;
        while (start < s.length() && set.indexOf(s.charAt(start)) != -1) {
            start++;
        }
        int end = s.length();
        while (end > start && set.indexOf(s.charAt(end - 1)) != -1) {
            end--;
        }
        return s.substring(start, end);
    }

    private static void check_utf8_MakeHexIfBinary(String s) {
        // Mock implementation
    }
}
